compiler/semantic/op.go

```diff
@@ -987,118 +870,117 @@ func (a *analyzer) semFuncDecls(decls []*ast.FuncDecl) ([]*dag.Func, error) {
+			a.error(d, err)
 		}
 	}
 	return a.semExpr(body)
 }
 
-func (a *analyzer) semOpDecl(d *ast.OpDecl) error {
+func (a *analyzer) semOpDecl(d *ast.OpDecl) {
 	m := make(map[string]bool)
 	for _, p := range d.Params {
 		if m[p] {
-			return fmt.Errorf("%s: duplicate parameter %q", d.Name, p)
+			a.error(d, fmt.Errorf("duplicate parameter %q", p))
+			a.scope.DefineAs(d.Name, &opDecl{bad: true})
+			return
 		}
 		m[p] = true
 	}
-	return a.scope.DefineAs(d.Name, &opDecl{ast: d, scope: a.scope})
+	if err := a.scope.DefineAs(d.Name, &opDecl{ast: d, scope: a.scope}); err != nil {
+		a.error(d, err)
+	}
 }
 
-func (a *analyzer) semVars(defs []ast.Def) ([]dag.Def, error) {
+func (a *analyzer) semVars(defs []ast.Def) []dag.Def {
 	var locals []dag.Def
 	for _, def := range defs {
-		e, err := a.semExpr(def.Expr)
-		if err != nil {
-			return nil, err
-		}
-		name := def.Name
-		if err := a.scope.DefineVar(name); err != nil {
-			return nil, err
+		e := a.semExpr(def.Expr)
+		if err := a.scope.DefineVar(def.Name); err != nil {
+			a.error(def, err)
+			continue
 		}
 		locals = append(locals, dag.Def{
-			Name: name,
+			Name: def.Name,
 			Expr: e,
 		})
 	}
-	return locals, nil
+	return locals
 }
 
-func (a *analyzer) semOpAssignment(p *ast.OpAssignment) (dag.Op, error) {
+func (a *analyzer) semOpAssignment(p *ast.OpAssignment) dag.Op {
 	var aggs, puts []dag.Assignment
-	for _, assign := range p.Assignments {
+	for _, astAssign := range p.Assignments {
 		// Parition assignments into agg vs. puts.
-		a, err := a.semAssignment(assign)
-		if err != nil {
-			return nil, err
-		}
-		if _, ok := a.RHS.(*dag.Agg); ok {
-			if _, ok := a.LHS.(*dag.This); !ok {
-				return nil, errors.New("summarize: aggregate output field must be static")
+		assign := a.semAssignment(astAssign)
+		if _, ok := assign.RHS.(*dag.Agg); ok {
+			if _, ok := assign.LHS.(*dag.This); !ok {
+				a.error(astAssign.LHS, errors.New("aggregate output field must be static"))
 			}
-			aggs = append(aggs, a)
+			aggs = append(aggs, assign)
 		} else {
-			puts = append(puts, a)
+			puts = append(puts, assign)
 		}
 	}
 	if len(puts) > 0 && len(aggs) > 0 {
-		return nil, errors.New("mix of aggregations and non-aggregations in assignment list")
+		a.error(p, errors.New("mix of aggregations and non-aggregations in assignment list"))
+		return badOp()
 	}
 	if len(puts) > 0 {
 		return &dag.Put{
 			Kind: "Put",
 			Args: puts,
-		}, nil
+		}
 	}
 	return &dag.Summarize{
 		Kind: "Summarize",
 		Aggs: aggs,
-	}, nil
+	}
 }
 
-func assignmentHasDynamicLHS(assignments []dag.Assignment) bool {
-	for _, a := range assignments {
-		if _, ok := a.LHS.(*dag.This); !ok {
+func (a *analyzer) checkStaticAssignment(asts []ast.Assignment, assignments []dag.Assignment) bool {
+	for k, assign := range assignments {
+		if _, ok := assign.LHS.(*dag.BadExpr); ok {
+			continue
+		}
+		if _, ok := assign.LHS.(*dag.This); !ok {
+			a.error(asts[k].LHS, errors.New("assignment key must be static"))
```

Comment:
Nit: Make this consistent with the similar error in semOpAssignment.

```suggestion
			a.error(asts[k].LHS, errors.New("output field must be static"))
```